src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsV2ManagedIdentitySource.cs

```diff
@@ -0,0 +1,156 @@
+
+        public static async Task<bool> GetCsrMetadataAsync(RequestContext requestContext)
+        {
+            string uami = null;
+            switch (requestContext.ServiceBundle.Config.ManagedIdentityId.IdType)
+            {
+                case AppConfig.ManagedIdentityIdType.ClientId:
+                    uami = requestContext.ServiceBundle.Config.ManagedIdentityId.UserAssignedId;
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 supports user-assigned client id.");
+                    break;
+
+                case AppConfig.ManagedIdentityIdType.ResourceId:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 doesn't support user-assigned resource id. Either provide a client id, or use a system-assigned managed identity.");
+                    return false;
+
+                case AppConfig.ManagedIdentityIdType.ObjectId:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 doesn't support user-assigned object id. Either provide a client id, or use a system-assigned managed identity.");
+                    return false;
+
+                default:
+                    requestContext.Logger.Info("[Managed Identity] ImdsV2 supports system-assigned managed identity.");
+                    break;
+            }
+
+            string queryParams = $"api-version={ImdsManagedIdentitySource.ImdsApiVersion}";
+            if (!string.IsNullOrEmpty(uami))
+            {
+                queryParams += $"&uaid={uami}";
+            }
+
+            Uri csrMetadataEndpoint = ImdsManagedIdentitySource.GetValidatedEndpoint(requestContext.Logger, CsrMetadataPath, queryParams);
+
+            var headers = new Dictionary<string, string>
+            {
+                { "Metadata", "true" },
+                { "x-ms-client-request-id", Guid.NewGuid().ToString() }
+            };
+
+            IRetryPolicyFactory retryPolicyFactory = requestContext.ServiceBundle.Config.RetryPolicyFactory;
+            IRetryPolicy retryPolicy = retryPolicyFactory.GetRetryPolicy(RequestType.ManagedIdentityDefault);
+
+            // CSR metadata GET request
+            HttpResponse response;
+
+            try
+            {
+                response = await requestContext.ServiceBundle.HttpManager.SendRequestAsync(
+                    csrMetadataEndpoint,
+                    headers,
+                    body: null,
+                    method: System.Net.Http.HttpMethod.Get,
+                    logger: requestContext.Logger,
+                    doNotThrow: false,
+                    mtlsCertificate: null,
+                    validateServerCertificate: null,
+                    cancellationToken: requestContext.UserCancellationToken,
+                    retryPolicy: retryPolicy)
+                .ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                // try/catch is for testing purposes, to avoid adding a mock for this request
```

Comment:
I would avoid as much as possible adding code to the src projects just for testing purposes if you have another option like adding a mock handler. Is there supposed to be an exception when this fails in production?

Reply:
this try/catch is actually being used now, now just for unit testing